Dynamic coating wall roof feature

yujiex opened the pull request:
Pull request overview
---------------------
## Introduction
Currently, EnergyPlus assumes constant thermal and solar absorptance for exterior wall and roof materials. In order to model dynamic coatings with changing solar or thermal absorptance, users need to use the EnergyManagementSystem (EMS) feature to write code to overwrite these properties. Although this can be done by advanced users, it can be a challenge for normal EnergyPlus users as learning and debugging EMS code is not an easy task.

The proposed new feature will enable user input of the stimulus-absorptance relationship in the form of a table or curve. The variables triggering solar or thermal absorptance change can be surface temperature, solar radiation, space heating/cooling status, a schedule, etc. An idd object will be added to capture detailed input of solar and thermal absorptance as a function of different stimuli.

The NFP document is in PR #9755 
[link to design document](https://github.com/NREL/EnergyPlus/blob/f9d3459053ed3e00ff913ca93050f2781d617966/design/FY2023/NFP-DynamicCoatingWallRoof.md)

## example case
The following figure shows a comparison between 1ZoneUncontrolled.idf and 1ZoneUncontrolled_variableThermalSolarAbs.idf, using the USA_CO_Golden-NREL.724666_TMY3.epw weather input
![design_day_cmp_1ZoneUncontrolled](https://user-images.githubusercontent.com/13421161/216721143-eacef2cb-9207-4178-8c71-dbdf70f05aec.png)

1ZoneUncontrolled_variableThermalSolarAbs.idf has a dynamic coating on Zn001:Wall001. The coating has variable thermal absorptance and solar absorptance based on surface temperature.
![thermal_solar_abs_temp_curve](https://user-images.githubusercontent.com/13421161/215914358-c7cbf92d-5ef0-417b-b3e1-d65590484617.png)

The following is a comparison among the feature implementation, the EMS approach, against a baseline no-coating case. For simplicity, the coating has only variable thermal absorptance. The calling point of the EMS approach is "BeginZoneTimestepBeforeInitHeatBalance". The feature and EMS approach have very similar surface temperature profiles. The maximum surface temperature difference between the feature and the EMS approach is 0.03C. This feature only overwrites the exterior surface absorptance, while EMS overwrites both the interior and the exterior surface absorptance. This seems to be the cause of the difference in the surface temperature output in this figure.
![design_day_cmp_1ZoneUncontrolled_ems_vs_feature](https://user-images.githubusercontent.com/13421161/218185418-c2a2b95d-cfa2-4eed-b5a7-776387493b77.png)

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

yujiex committed:
Idd change

yujiex committed:
Initial commit of testfile 1ZoneUncontrolled

yujiex committed:
initial implementation and simple test file

yujiex committed:
default constructor -> init at declaration in MaterialProperties

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoofFeature

yujiex committed:
add test idf to CMakeLists

yujiex committed:
clang-format

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -2758,4 +2763,216 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
+    GetVariableSolarAbsorptanceInput(state, errorsFound);
+}
+
+// assume only one matching surface
```
This is my biggest concern about the implementation. Where does this assumption come from? Why is it valid? And even if it is, why is the chosen surface the first one that matches?

The answers to these questions aside, the implementation here is "inverted". Materials should not reference Zones or Surfaces, that is bad design. You don't want that circularity either in the code (notice "Surfaces have to be set up before this can be set up" and also the large number of otherwise header files you have to include in `Material.cc`) or at runtime. Even if only one Surface needs to know about these new properties, the API for activating this functionality should be attached to the Surface, not to the Material. 

FWIW I notice that the EMS implementation of this feature has the same problem, i.e., it is attached to the material as opposed to each instance of the material. The Material class should be stateless, because it is presumably shared by many Surface instances, each of which will require its own material state.

yujiex replied:
The actual product is a material but the thermal solar absorption for this dynamic coating material is driven by surface properties or zone properties. So is it better to have this coating be added to surfaces properties rather than material?

amirroth replied:
You have to split it up. The state needs to go with Zone and/or Surface, the non-state stuff that is shared by all Surfaces and Zones (ControlSignal, Function, and Schedule) can stay with the Material class.

The more I look at it, there is a lot of this type of confusion in Material and probably elsewhere in EnergyPlus also.

yujiex replied:
Now I think I'm clear. When I apply the overwrite, it will be at the surface/zone level, rather than the material level.

amirroth replied:
Correct. That way each surface gets its own absorptance based on its own temperature and/or the temperature of the zone it is connected to.

amirroth replied:
And you can apply this to all surfaces that use this material rather than just the first one (which is probably just as important).

yujiex replied:
exactly, with the previous design, the user will have to define a different material for each surface, which is the current EMS way. I got confused and followed the EMS way. I should definitely move the state out of the material and make surface reference material.

amirroth commented on idd/Energy+.idd.in:
```diff
@@ -7604,6 +7604,54 @@ MaterialProperty:VariableThermalConductivity,
        \units W/m-K
        \type real
 
+MaterialProperty:VariableThermalAbsorptance,
```
General question. Are `MaterialProperty` orthogonal? Can a `Material` have any combination of them?

yujiex replied:
I believe for MaterialProperty:x, there can only be at most one of each kind. There can be MaterialProperty:a, MaterialProperty:b, MaterialProperty:c, etc. for the same material where a, b, c are different add-ons. For example, the `MaterialProperty:PhaseChange` object has just one "Name" field pointing to the material object it references. There is not a name for the add-on itself. So one material cannot have multiple `MaterialProperty:PhaseChange` objects as they will not be distinguishable. 

The two I added do have their own names in addition to the material they reference, but I still only allow for one of each kind for each material as other add-ons.

amirroth replied:
That part I get. I meant are the different `MaterialProperty` objects orthogonal to each other relative to a single `Material` object. Can a single `Material` have `MaterialProperty:PhaseChange` and `MaterialProperty:VariableThermalAbsorptance` and `MaterialProperty:MoisturePenetrationDepth` and the rest of them?

yujiex replied:
These are the MaterialProperty: ones. From the names it looks like they should be orthogonal (although need to double check the PhaseChange and PhaseChangeHysterisis)
<img width="437" alt="image" src="https://user-images.githubusercontent.com/13421161/215599581-9849701c-3c7d-4cef-98bb-c52dca88ddd2.png">

mjwitte replied:
Some of these are mutually exclusive for any given surface. It's actually kind of messy. The two new ones, however can work in combination with any of the others.

MaterialProperty:MoisturePenetrationDepth:Settings - requires HeatBalanceAlgorithm=MoisturePenetrationDepthConductionTransferFunction

MaterialProperty:PhaseChange and MaterialProperty:PhaseChangeHysteresis require HeatBalanceAlgorithm = CondFD (ConductionFiniteDifference), but you'd only use one or the other.

MaterialProperty:VariableThermalConductivity requires HeatBalanceAlgorithm = CondFD Can be used together with one of MaterialProperty:PhaseChange*

MaterialProperty:HeatAndMoistureTransfer:* are user as a group and require HeatBalanceAlgorithm = CombinedHeatAndMoistureFiniteElement

MaterialProperty:GlazingSpectralData only applies to windows, all the others only apply to opaque surfaces.

So, not very orthogonal, kind of skewed or twisted.

yujiex replied:
Thanks for the clarification @mjwitte. Glad the new ones can work together with the rest add-ons.

amirroth replied:
Oh what a tangled web we weave ...

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -2758,4 +2763,216 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
+    return -1;
+}
+
+Real64 GetMaterialOverrideValue(EnergyPlusData &state, const VariableAbsAddOn &thisAbsAddOn)
```
Essentially, this function and the next one need to be based on Surface, rather than on Material. You should make a list of the indices of all exterior surfaces that have these types of materials. And then at the correct point in the heat balance calculation, loop over those surfaces to grab the new absorptance if needed. You probably also need a per-surface absorptance array that the normal calculations use and that can this loop can modify. And the code should probably move to `HeatBalanceSurfaceManager`.

yujiex replied:
Indeed. I will keep an array of surface indices with dynamic coating and retrieve the control signal variable value and the overwritten value at the surface heat balance calculation.

yujiex committed:
update comments of SurfAbs*, inside outside reversed

yujiex committed:
Change override to surface level

yujiex committed:
clang-format

yujiex committed:
fix headers after moving MaterialAddOnOverride

amirroth commented on src/EnergyPlus/Material.hh:
```diff
@@ -132,239 +132,243 @@ namespace Material {
-              GlassSpecAngFRefleDataPtr(0), GlassSpecAngBRefleDataPtr(0)
-        {
-        }
+    struct VariableAbsAddOn
```
I also forgot to mention, but ... you should get rid of this additional struct and put these into the `MaterialProperties` struct (they will end up in a child class after the hierarchy refactor). As a result, you should not need to store the first three fields. You should really only need `variableAbsorptanceCtrlSignal`, `variableAbsporptanceScheduleIdx`, and `variableAbsorptanceFunctionIdx`.

yujiex replied:
Oh, yeah. I was wondering before whether add-on's should be in separate struct. I saw MaterialProperty:VariableThermalConductivity is in `state.dataHeatBalFiniteDiffMgr->MaterialFD` and MaterialProperty:GlazingSpectralData is in `state.dataHeatBal->SpectralData`

amirroth replied:
We will consolidate all of those (and Blind and Screen and PhaseChangeMaterial and a few other things) into the hierarchy as well. Maybe not as part of this refactoring effort, but certainly later.

amirroth replied:
In general, if the established pattern in EnergyPlus seems wrong or inefficient to you then don't follow it blindly. Ask someone (me or Edwin or Mike or anyone) what they would suggest or if you have a better idea on your own, just do that. Many of the problems in EnergyPlus started as one doing something kind of weird, and then other people just copying it.

yujiex replied:
I'll ask around next time.

yujiex committed:
use global surfList for addon surfs, remove VariableAbsAddOn struct

yujiex committed:
Fix quotation marks in added idd objects

amirroth commented on idd/Energy+.idd.in:
```diff
@@ -7604,6 +7604,54 @@ MaterialProperty:VariableThermalConductivity,
+  A4 , \field Trigger Thermal Absorptance Function Name
+       \note A Curve:* or Table:Lookup object encoding the relationship between the trigger and surface thermal absorptance.
+  A5 ; \field Thermal Absorptance Schedule Name
+       \note only used when Control Signal = "Scheduled"
```
Do these materials commonly have both variable thermal and variable solar absorptance or just or the other? I'm wondering if these should be combined into a single object... 🤔

amirroth commented on src/EnergyPlus/DataSurfaces.hh:
```diff
@@ -1382,6 +1382,8 @@ struct SurfacesData : BaseGlobalStruct
     std::vector<int> AllExtSolWinWithFrameSurfaceList; // List of all exterior solar window surfaces with a frame and divider
     std::vector<int> AllHTKivaSurfaceList;             // List of all Kiva foundation surfaces
     std::vector<int> AllSurfaceListReportOrder;        // List of all surfaces - output reporting order
+    std::vector<int> AllVaryThermalAbsOpaqSurfaceList; // List of all opaque exterior surfaces with dynamic coating
```
Yes, this is what I meant. :+1:

yujiex replied:
:)

amirroth replied:
This could become a single list.

yujiex replied:
right, since we combined the idd objects

amirroth commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -328,6 +328,11 @@ namespace HeatBalanceManager {
         GetBuildingData(state, ErrorsFound); // Read building data from input file
 
+        // need surface data to be set up before these add-on can be constructed
+        Material::GetMaterialAddOnInput(state, ErrorsFound); // Read variable thermal and solar absorptance addon data
```
Can probably just add this code to `GetMaterialData` without this separate function, no?

yujiex replied:
good point. I'll move it into `GetMaterialData`.

amirroth commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -328,6 +328,11 @@ namespace HeatBalanceManager {
+        // need surface data to be set up before these add-on can be constructed
+        Material::GetMaterialAddOnInput(state, ErrorsFound); // Read variable thermal and solar absorptance addon data
+
+        Material::GetAddOnOverrideSurfaceList(state);
```
This function should be part of `Surface`, not `Material`. Should not include any surface stuff in the Material module.

yujiex replied:
Moved to HeatBalanceSurfaceManager

yujiex replied:
well maybe move to DataSurfaces

amirroth commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -247,6 +249,52 @@ void ManageSurfaceHeatBalance(EnergyPlusData &state)
+
+void MaterialAddOnOverride(EnergyPlusData &state)
+{
+    for (int surfNum : state.dataSurface->AllVaryThermalAbsOpaqSurfaceList) {
```
Is it possible for the same surface to appear in both lists? In that case, the override value from one would overwrite the override value from the other, no?

amirroth commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -356,6 +404,9 @@ void InitSurfaceHeatBalance(EnergyPlusData &state)
     }
 
+    // yujie: variable thermal solar absorptance overrides
+    MaterialAddOnOverride(state);
```
Not a descriptive name for a the function. Maybe `UpdateVariableAbsorptances` would be better. Currently, the function name is not telling you what about the material is being overridden.

yujiex replied:
This is indeed a more descriptive name

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -2758,4 +2762,130 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
+    GetVariableSolarAbsorptanceInput(state, errorsFound);
+}
+
+int GetMaterialNumFromSurfNum(EnergyPlusData &state, int surfNum)
```
Move these functions into the `DataSurfaces` module. That way you won't have to include `Donstruction.hh` and `DataSurfaces.hh` here.

yujiex replied:
moved it to DataSurfaces

amirroth commented on idd/Energy+.idd.in:
```diff
@@ -7604,6 +7604,54 @@ MaterialProperty:VariableThermalConductivity,
+       \note A Curve:* or Table:Lookup object encoding the relationship between the trigger and surface solar absorptance.
+  A5 ; \field Solar Absorptance Schedule Name
+       \note only used when Control Signal = "Scheduled"
+
```
Based on our discussion yesterday, I think it makes sense to combine these into a single `MaterialProperties:VariableAbsorptance` object with a single control signal and separate functions and schedules. If the function name is empty, it would mean that the absorptance component is not modulated.

yujiex replied:
make sense. Will combine them into one field with separate function and schedule for thermal and solar

yujiex committed:
move surf functions to DataSurfaces, rename and combine functions

- move GetAddOnOverrideSurfaceList and GetMaterialNumFromSurfNum into DataSurfaces
- move GetMaterialAddOnInput into GetMaterialData
- rename function to UpdateVariableAbsorptances

yujiex committed:
Combine solar thermal absorptance to one idd object

yujiex committed:
clang format

yujiex committed:
remove unused functions from header, fix comments

amirroth commented on idd/Energy+.idd.in:
```diff
@@ -7604,6 +7604,58 @@ MaterialProperty:VariableThermalConductivity,
+       \key SpaceHeatingCoolingMode
+       \key Scheduled
+       \default SurfaceTemperature
+  A4 , \field Trigger Thermal Absorptance Function Name
```
What is the word "Trigger" referring to here?

yujiex replied:
It refers to the control signal, I'll change it to "Control Signal-Thermal Absorptance Function Name"

amirroth replied:
You can just delete it also. It’s not necessary.

yujiex replied:
oh that works too. Maybe I'll just delete it.

amirroth commented on src/EnergyPlus/DataSurfaces.cc:
```diff
@@ -734,4 +734,26 @@ Real64 AbsBackSide(EnergyPlusData &state, int SurfNum)
     return AbsorptanceFromExteriorBackSide + AbsorptanceFromInteriorBackSide;
 }
 
+void GetAddOnOverrideSurfaceList(EnergyPlusData &state)
```
Probably should rename to `GetVariableAbsorptanceSurfaceList`?

yujiex replied:
right, that's better

amirroth commented on src/EnergyPlus/DataSurfaces.cc:
```diff
@@ -734,4 +734,26 @@ Real64 AbsBackSide(EnergyPlusData &state, int SurfNum)
+void GetAddOnOverrideSurfaceList(EnergyPlusData &state)
+{
+    for (int surfNum = 1; surfNum <= state.dataSurface->TotSurfaces; ++surfNum) {
+        int materNum = GetMaterialNumFromSurfNum(state, surfNum);
```
Breaking this out into its own function is not really helping you here. Would be better to just inline the logic.

yujiex replied:
yeah, I'll just inline the content. Although I'm not sure if this type of lookup operation is performed a lot, whether it would be better to make a helper function like this.

amirroth commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -247,6 +249,67 @@ void ManageSurfaceHeatBalance(EnergyPlusData &state)
 // Beginning Initialization Section of the Module
 //******************************************************************************
 
+Real64 truncateOverrideValue(Real64 value)
```
Breaking these things out into functions is not actually helping, it's just making you pull out and pass parameters around.

yujiex replied:
Removed the function and inlined the content

amirroth commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -247,6 +249,67 @@ void ManageSurfaceHeatBalance(EnergyPlusData &state)
+    Real64 overrideValueThermal = 0.0;
+    Real64 overrideValueSolar = 0.0;
+    if (controlSignal == Material::VariableAbsCtrlSignal::Scheduled) {
+        overrideValueThermal = ScheduleManager::GetCurrentScheduleValue(state, scheduleIdxThermal);
```
One of these could be `-1`, no? Don't you have to check if they are `-1` the same way you are checking the function indices below?

yujiex replied:
The scheduleIdx will be 0 if GetScheduleIndex didn't find it.
Then the GetCurrentScheduleValue will return 0 if the input index of the schedule is 0. 
So maybe the -1 case won't appear?

yujiex replied:
Oh I initialized them in the struct as -1

amirroth replied:
Right. I meant one of them could be 0. Shouldn’t you check for that?

yujiex replied:
If the schedule index is 0, then the returned override value will also be 0. The following chunk will make sure to overwrite only when the value is positive
```
        std::pair<Real64, Real64> overrideValues = GetSurfAddOnOverrideValue(state,
                                                                             surfNum,
                                                                             thisMaterial->absorpVarCtrlSignal,
                                                                             thisMaterial->absorpThermalVarSchedIdx,
                                                                             thisMaterial->absorpThermalVarFuncIdx,
                                                                             thisMaterial->absorpSolarVarSchedIdx,
                                                                             thisMaterial->absorpSolarVarFuncIdx);
        if (overrideValues.first > 0.0) state.dataHeatBalSurf->SurfAbsThermalExt(surfNum) = overrideValues.first;
        if (overrideValues.second > 0.0) state.dataHeatBalSurf->SurfAbsSolarExt(surfNum) = overrideValues.second;
```

amirroth replied:
Ok. I understand that you didn't create this behavior where if a schedule doesn't exist then you can still get a value out of it but that value is just zero, but this behavior is terrible, opaque, and leads to errors. Look at the gymnastics you have to do to deal with it, you have to clamp the value to 0.00001 and 0.99999 and then check that it's not 0.0 before you set it, and that logic is distributed over three different places. It's easier and clearer to handle this condition explicitly.

```cpp
if (thisMaterial->schedIdxSolar > 0) 
   state.dataHeatBalSurf->SurfAbsSolarExt(SurfNum) = std::clamp(GetCurrentScheduleValue(state, thisMaterial->schedIdxSolar), 0.0, 1.0);
```

yujiex replied:
Oh! this way is nicer

amirroth commented on src/EnergyPlus/HeatBalanceSurfaceManager.hh:
```diff
@@ -55,6 +55,7 @@
 #include <EnergyPlus/Data/BaseData.hh>
 #include <EnergyPlus/DataWindowEquivalentLayer.hh>
 #include <EnergyPlus/EnergyPlus.hh>
+#include <EnergyPlus/Material.hh>
```
Should not need this if you eliminate the "helper" functions.

yujiex replied:
removed it

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -2756,6 +2758,99 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
         m->phaseChange = HysteresisPhaseChange::HysteresisPhaseChange::factory(state, m->Name);
     }
+
+    GetMaterialAddOnInput(state, ErrorsFound); // Read variable thermal and solar absorptance add-on data
```
`GetVariableAbsorptanceInput`?

yujiex replied:
right, more descriptive this way

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -2756,6 +2758,99 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
+    GetMaterialAddOnInput(state, ErrorsFound); // Read variable thermal and solar absorptance add-on data
+}
+
+void ReadingOneMaterialAddOn(EnergyPlusData &state, int i, bool &errorsFound)
```
Again, "helper" functions like this don't actually help. They just add overhead because you have to construct and allocate the parameter arrays for each call instead of just once.

yujiex replied:
yeah, previously I was thinking since the thermal and solar one mostly share the same reading procedure, there should be such a helper. Now that they are combined into one idd object, there's no need for this I guess.

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -2756,6 +2758,99 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
+    Material::VariableAbsCtrlSignal controlSignal = Material::VariableAbsCtrlSignal::SurfaceTemperature; // default value
+    controlSignal = static_cast<VariableAbsCtrlSignal>(getEnumerationValue(VariableAbsCtrlSignalUC, UtilityRoutines::MakeUPPERCase(alphas(3))));
+    //    init to 0 as GetScheduleIndex returns 0 for not-found schedule or curves
+    int functionIdxThermal = 0;
```
You may as well just use `thisMaterial->variableThermalAbsorptanceFunctionIdx` instead of the temporary here.

yujiex replied:
right. changed it to `thisMaterial->variableThermalAbsorptanceFunctionIdx`

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -2756,6 +2758,99 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
+    if (controlSignal == VariableAbsCtrlSignal::Scheduled) {
+        scheduleIdxThermal = ScheduleManager::GetScheduleIndex(state, alphas(5));
+        scheduleIdxSolar = ScheduleManager::GetScheduleIndex(state, alphas(7));
+        if (scheduleIdxThermal + scheduleIdxSolar == 0) {
```
Don't replace logic with math. It's not robust and also confusing to the reader. `if (schedIdxThermal == 0 && schedIdxSolar == 0)` is better.

yujiex replied:
good point. change it to two equality checks

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -2756,6 +2758,99 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
+}
+
+void GetMaterialAddOnInput(EnergyPlusData &state, bool &errorsFound)
+{
```
This function is doing nothing except calling the other function in a loop. Splitting into two functions doesn't really make sense here.

yujiex replied:
combined the ReadingOneMaterialAddOn into it.

amirroth commented on src/EnergyPlus/Material.hh:
```diff
@@ -132,232 +132,224 @@ namespace Material {
+        bool AbsorpThermalEMSOverrideOn = false; // if true, then EMS calling to override value for thermal absorptance
+        Real64 AbsorpThermalEMSOverride = 0.0;   // value to use when EMS calling to override value for thermal absorptance
+        // dynamic thermal and solar absorptance coating parameters
+        VariableAbsCtrlSignal variableAbsorptanceCtrlSignal = VariableAbsCtrlSignal::Invalid;
```
Ok, so all of the variables here are called `AbsorpSolar` and `AbsorpThermal`.  So maybe rename new variables to match, e.g., `absorpVarCtrlSignal`, `absorpSolarVarSchedIdx`, `absorpThermalVarFuncIdx`, etc. This way, the reader knows all of these variables go together.

yujiex replied:
renamed them. look more consistent this way

yujiex committed:
renaming, combining functions

amirroth commented on src/EnergyPlus/Material.cc:
```diff
@@ -2756,6 +2758,85 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
+            return;
+        }
+
+        Material::VariableAbsCtrlSignal controlSignal = Material::VariableAbsCtrlSignal::SurfaceTemperature; // default value
```
Shouldn't this be `thisMaterial->absorpVarCtrlSignal =`? Don't you have to save this in `thisMaterial`?

yujiex replied:
yes, I do need to save it. forgot to add it back after removing temp variables

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoofFeature

yujiex committed:
fix control signal field not stored

yujiex committed:
add back solar absorptance curve in test idf

yujiex committed:
Make the clamp to 0.0001 and 0.9999 based on idx > 0

yujiex committed:
fix undeclared local var after combining functions

yujiex committed:
unit test initial commit

yujiex committed:
change EXPECT_EQ() to EXPECT_TRUE(compare_enums()) for enum compare

yujiex committed:
update documentation

yujiex committed:
add 2 warning msg, add err msg check in unit test, fix fieldName idx

yujiex committed:
Update documentation regarding EMS vs feature difference

yujiex committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoofFeature

yujiex committed:
Documentation and warning msg on feature and EMS difference

yujiex committed:
Fix warning arg count err

mjwitte commented on src/EnergyPlus/Material.cc:
```diff
@@ -2756,6 +2758,106 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
+            }
+            if ((thisMaterial->absorpThermalVarSchedIdx > 0) || (thisMaterial->absorpSolarVarSchedIdx > 0)) {
+                ShowWarningError(state,
+                                 format("{}: Non-scedule control signal is chosen. Thermal or solar absorptance schedule name is going to be "
```
`Non-schedule ...`

yujiex replied:
Thanks. fixed it.

mjwitte commented on src/EnergyPlus/DataSurfaces.cc:
```diff
@@ -734,4 +734,34 @@ Real64 AbsBackSide(EnergyPlusData &state, int SurfNum)
     return AbsorptanceFromExteriorBackSide + AbsorptanceFromInteriorBackSide;
 }
 
+void GetVariableAbsorptanceSurfaceList(EnergyPlusData &state)
```
If I am following correctly, this function will walk through every material layer of every surface even if there are zero `MaterialProperty:VariableAbsorptance` objects in the input. I would add a global flag `AnyVariableAbsorptance`, similar to `DataHeatBalance::AnyAirBoundary`, default to false, and set true in `GetVariableAbsorptanceInput`. Then this function can return immediately if `!AnyVariableAbsorptance`.

yujiex replied:
Indeed. I'll add a global flag to short-circuit it when no MaterialProperty:VariableAbsorptance is present.

yujiex committed:
fix documentation duplicate label

mjwitte commented on src/EnergyPlus/DataSurfaces.cc:
```diff
@@ -734,4 +734,34 @@ Real64 AbsBackSide(EnergyPlusData &state, int SurfNum)
 
+void GetVariableAbsorptanceSurfaceList(EnergyPlusData &state)
+{
+    for (int surfNum = 1; surfNum <= state.dataSurface->TotSurfaces; ++surfNum) {
```
This can loop over `DataSurfaces::AllHTSurfaceList`, then you don't need to check for `ConstrNum <=0`. All HT surfaces have a valid construction.

yujiex replied:
oh nice. I'll change it to `DataSurfaces::AllHTSurfaceList`

mjwitte commented on src/EnergyPlus/DataSurfaces.cc:
```diff
@@ -734,4 +734,34 @@ Real64 AbsBackSide(EnergyPlusData &state, int SurfNum)
+                state.dataSurface->AllVaryAbsOpaqSurfaceList.push_back(surfNum);
+            }
+        }
+        // check for dynamic coating defined on the non-outside layer of a construction
```
This check can be done more efficiently by looping over all constructions (separate from the surface loop) and just throwing an error once per construction and mention the construction name.  For both warnings, I would add something like "This will be ignored."

yujiex replied:
right right. Shouldn't nest it in the surface loop.

yujiex committed:
fix typo, add short-circuit, rearrange warning loop, mod err msg

yujiex committed:
auto const &thisSurface, to prevent copying

amirroth commented on src/EnergyPlus/DataSurfaces.cc:
```diff
@@ -734,4 +734,43 @@ Real64 AbsBackSide(EnergyPlusData &state, int SurfNum)
+{
+    if (!state.dataHeatBal->AnyVariableAbsorptance) return;
+    for (int surfNum : state.dataSurface->AllHTSurfaceList) {
+        auto const thisSurface = state.dataSurface->Surface(surfNum);
```
This is why we need to eliminate all scalar uses of `auto`. So that we can check automatically that all non-iterator/non-lambda uses of `auto` are `auto &`.

yujiex replied:
indeed. Just fixed this instance. I'll remember `&` in the future.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -7604,6 +7604,61 @@ MaterialProperty:VariableThermalConductivity,
+  A7 ; \field Solar Absorptance Schedule Name
+       \note only used when Control Signal = "Scheduled"
+
+MaterialProperty:VariableSolarAbsorptance,
```
@yujiex This object should be deleted, correct?

yujiex replied:
Oh, I forgot to take that out of idd. We only have one object now, MaterialProperty:VariableAbsorptance

yujiex replied:
just removed it

Myoldmopar replied:
Was this comment from an old object?  It is showing up on the `MaterialProperty:HeatAndMoistureTransfer:Settings` object now, but I don't think that object would be deleted here.

yujiex committed:
Remove unused idd object

yujiex committed:
Modify design doc to only have MaterialPropery:VariableAbsorptance

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into dynamicCoatingWallRoofFeature

Myoldmopar commented on doc/ems-application-guide/src/ems-actuators/thermal-envelope.tex:
```diff
@@ -43,7 +43,18 @@ \subsection{Surface Convection Heat Transfer Coefficient}\label{surface-convecti
+paints. Note that for a single-layer construction, both the inside and outside
+properties will be overwritten. For overwriting outside properties alone, please
+use ``MaterialProperty:VariableAbsorptance'' (see Section
+\ref{materialpropertyvariableabsorptance}).
```
Github is complaining about this reference being undefined.  I see it defined below in the IO reference, but that's a different document than this EMS reference paragraph.  I assume this link should be removed here.

Myoldmopar commented on src/EnergyPlus/Material.cc:
```diff
@@ -2756,6 +2758,107 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
+            errorsFound = true;
+            return;
+        }
+        auto *thisMaterial = state.dataMaterial->Material(MaterNum);
```
Auto with pointer feels a little vague to me.  I'd like to minimize this pattern in the future.  I think it would be easy enough to be explicit about the type here, it's not like it's a long iterator typename.  This is not critical for merge today.

amirroth replied:
What? Why? This isn't an iterator, it's no different than `auto &`. The reason we are using `*` here instead of `&` is that we changed the Material array to be an array of `Material` pointers rather than `Material` objects so that we can take advantage of inheritance (we're not doing it yet, but it is coming).

Myoldmopar replied:
My issue with it is because of what auto means when you are assigning to a pointer object.  In this case, both of these are the same thing:

```
auto * thisMaterial = state.dataMaterial->Material(MaterNum);
auto thisMaterial = state.dataMaterial->Material(MaterNum);
```

They both result in thisMaterial being a pointer to a material.  I'm not upset about there being a pointer, I just dislike that auto sneaks in the pointer automatically, meaning...you don't need the * anymore.  Whereas if you use the typename, you get:

```
MaterialProperties * thisMaterial = state.dataMaterial->Material(MaterNum);
MaterialProperties thisMaterial = state.dataMaterial->Material(MaterNum);  << COMPILER ERROR
```

It's a small change to make things super clear.  I just don't like that `auto *` can ignore the * if it's already a pointer type.  Like...what if you actually wanted a pointer to a pointer and didn't want the compiler to ignore the *?

This is extremely far from a hill I would die on, so feel free to > /dev/null as needed.

Demo: https://godbolt.org/z/jhjjPb4ob

amirroth replied:
I would think about consenting to this if we were not using `auto &` so extensively. I still wouldn't consent to it, but I would think about it. 

We need to severely limit the ways in which we use "plain" `auto`. A good number of current uses of "plain" `auto` should be `auto &` to prevent unnecessary deep copies. If we use `auto thisMaterial = state.dataMaterial->Material(MaterialNum);` then it looks like one of those bad instances which we are trying to hunt down and convert to `auto &`. 

If you're starting to get the feeling that `auto` is a nightmare that is much more trouble than it's worth, then you're onto something.

Myoldmopar reviewed (approved):
I'll mark it as approved, but I am going to push a commit with a couple small tweaks prior to merge.  As of now it is still passing all unit and regression tests locally.

Myoldmopar committed:
Fix auto & usage, fix doc warning, fix return value warning from a separate PR

Myoldmopar commented:
@yujiex thanks for this.  I tossed a small commit on the end there with a couple fixes (one build warning that snuck into develop separately).  GitHub actions is complaining about using Ubuntu 18.04, which is in a brownout today as they try to get people like us to move away from it.  No need to wait for anything here though.  I'll merge this and then if something pops up I'll deal with it later.  Thanks to all the reviewers here as well.  Merging.

yujiex commented:
@Myoldmopar Thanks for fixing these. I'll remember to spell out things like MaterialProperties pointer in the future. The comment about deleting an idd object was about the VariableSolarAbsorptance object, which was deleted already. Thanks for catching the documentation reference issue too :)
